google/cloud/storagecontrol/v2/samples/storage_control_folder_samples.cc

```diff
@@ -0,0 +1,218 @@
+                std::string{"/folders/"} + source_folder_id;
+    // Start a rename operation and block until it completes. Real applications
+    // may want to setup a callback, wait on a coroutine, or poll until it
+    // completes.
+    auto renamed = client.RenameFolder(name, dest_folder_id).get();
+    if (!renamed) throw std::move(renamed).status();
+
+    std::cout << "Renamed: " << source_folder_id << " to: " << dest_folder_id
+              << "\n";
+  }
+  // [END storage_control_rename_folder]
+  (std::move(client), argv.at(0), argv.at(1), argv.at(2));
+}
+
+void AutoRun(std::vector<std::string> const& argv) {
+  namespace examples = google::cloud::testing_util;
+  namespace storagecontrol = google::cloud::storagecontrol_v2;
+  if (!argv.empty()) throw examples::Usage{"auto"};
+  examples::CheckEnvironmentVariablesAreSet(
+      {"GOOGLE_CLOUD_CPP_STORAGE_TEST_FOLDER_BUCKET_NAME"});
+  auto const bucket_name =
+      google::cloud::internal::GetEnv(
+          "GOOGLE_CLOUD_CPP_STORAGE_TEST_FOLDER_BUCKET_NAME")
+          .value();
+
+  auto client = storagecontrol::StorageControlClient(
+      storagecontrol::MakeStorageControlConnection());
+  auto generator = google::cloud::internal::DefaultPRNG(std::random_device{}());
+  auto const prefix = "storage-control-samples";
+  auto const folder_id = prefix + std::string{"-"} +
+                         google::cloud::internal::Sample(
+                             generator, 32, "abcdefghijklmnopqrstuvwxyz");
+  auto const dest_folder_id = prefix + std::string{"-"} +
+                              google::cloud::internal::Sample(
+                                  generator, 32, "abcdefghijklmnopqrstuvwxyz");
+  auto const create_time_limit =
+      std::chrono::system_clock::now() - std::chrono::hours(48);
+  // This is the only example that cleans up stale folders. The examples run in
+  // parallel (within a build and across the builds), having multiple examples
+  // doing the same cleanup is probably more trouble than it is worth.
+  [](google::cloud::storagecontrol_v2::StorageControlClient client,
+     std::string const& bucket_name, std::string const& prefix,
+     std::chrono::system_clock::time_point created_time_limit)
+      -> google::cloud::Status {
+    std::cout << "\nRemoving stale folders for examples" << std::endl;
+    std::regex re(prefix + R"re(-[a-z]{32})re");
+    auto const parent = std::string{"projects/_/buckets/"} + bucket_name;
+    for (auto folder : client.ListFolders(parent)) {
+      if (!folder) return std::move(folder).status();
+      if (!std::regex_match(folder->name(), re)) continue;
+      auto const create_time =
+          google::cloud::internal::ToChronoTimePoint(folder->create_time());
+      if (create_time > created_time_limit) continue;
+      (void)client.DeleteFolder(folder->name());
+    }
+    return {};
+  }(std::move(client), bucket_name, prefix, create_time_limit);
```

Comment:
Calling `std::move(client)` here invalidates the client and causes the crash.

Reply:
You are using the move-constructor: note that `std::move()` by itself just says "please use the move constructor if available".  There is a move constructor available: the compiler generated one, which performance member-by-member move construction. Oh, and because your lambda (now a function) consumes the `client` parameter by-value the constructor is called.

Move constructors (as the name implies) move or steal the contents of a the source object into the destination object [^1].  That leaves the source object in an undetermined state [^2].  Calling member functions that have pre-conditions may result in UB (undefined behavior).  In this case the contents are a `std::shared_ptr<>`, the shared pointer move constructor leaves the source with a `nullptr` and dereferencing a `nullptr` causes a crash.


[^1]: Well, move constructors can do whatever they want.  They can behave like copy constructors.  But *by convention* they move / steal the contents.
[^2]: Again, technically move constructors are required to leave the source object in a "consistent but undetermined state".  It could be that the source object is left in a state where it is usable, but *in general* one should assume move constructors leave you with an object where only functions without preconditions (e.g. assigning to the object) work.